### DISCUSSION TITLE
Optional separate active split border color

### DISCUSSION THREAD

**Discussion by yetyetanotherusername** [ghostty-org/ghostty#2727] [Feature Requests, Ideas]
All other tiling terminal emulators I have used so far (tilix & kitty) and also tmux offer the ability to indicate the active split via a separate border color, the inner half of the active split border can be set to a different, user defined color. Personally I prefer this over using opacity of inactive splits because I find the reduced opacity somewhat of a strain on my eyes. 

This could be controlled by new parameter in the settings, that, if set to a hex color code, would activate this feature.

Provided there is interest by the maintainer and provided this seems feasible, I would be interested in contributing this myself. 

Full disclosure, I have zero experience in zig, I do have experience in C++ though. This would probably take me some time to get my bearings on the codebase, so I would be completely fine with being sniped if someone more experienced feels like they want to and could do this quickly.

**Comment by mitchellh** [comment]
I'm good with this in general. Let's talk more specifically how this would look and act. Do you have a screenshot of another terminal that you're trying to emulate? I can imagine it in my head but just curious what you're looking at.

In terms of configuration, same thing: do you have links to docs for the other terminals?

Thanks!

**Comment by yetyetanotherusername** [comment]
Thanks for getting back,

this is what it looks like in kitty, along with a closeup so it's less ambiguous what is happening exactly. Essentially the inner half of every split border surrounding the active split turns a different color.

<img src="https://github.com/user-attachments/assets/f6ed6580-303d-4036-86ab-29de9a22211c" width="200" height="200" />
<img src="https://github.com/user-attachments/assets/bcc3eebe-ffa5-4aeb-b9e7-a3fe2852f3fd" width="200" height="200" />

Regarding config and again turning to kitty, there is `active_border_color` and `inactive_border_color`. kitty allows setting them to `none`, which is synonymous with having no visible borders (not sure how essential this is since the user can always set either to the background color).

For reference, this config is what produced the above screenshots.
```
active_border_color #4799ff
inactive_border_color #5e5c64
```

Documentation can be found [here](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.active_border_color) and [here](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.inactive_border_color).

tmux has separate [styles](https://man.openbsd.org/OpenBSD-current/man1/tmux.1#STYLES) for [active](https://man.openbsd.org/OpenBSD-current/man1/tmux.1#pane-active-border-style) and [inactive](https://man.openbsd.org/OpenBSD-current/man1/tmux.1#pane-border-style) border. This is a very powerful system that allows also setting background colors in a configurable margin around the border and tons of other stuff. However for the purpose of this proposal, I consider that way out of scope.

**Comment by jcollie** [comment]
On Linux, the split borders are drawn by GTK so it would take some advanced CSS hackery to have different colors depending on focus. So to make this work you'd probably need to make the GTK splits invisible (they still need to be there so that you can use the mouse to resize splits). Then the active/inactive border would need to be drawn by OpenGL/Metal.

**Reply by jcollie** [reply]
I'm in favor of the general idea though.

**Reply by yetyetanotherusername** [reply]
@jcollie Thank you for the insights. That sounds pretty involved. Can the line segments be styled independently? I'm thinking that at usual border width, it is very hard to notice that there are two colors anyway (see first of my screenshots). Could this act, if not as the fully fleshed out feature, then at least as a stop-gap?

**Reply by jcollie** [reply]
Changing the color of the separators is pretty easy, changing the colors of specific separators would be much more difficult as I don't think that there's a good/easy way to distinguish the separators without traversing the GTK widget tree and assigning some useful CSS classes/names to the widgets.

**Reply by yetyetanotherusername** [reply]
Yeah that makes sense. If that's ok with everybody, I'll make a branch and see if I can work my way towards a Linux POC of the approach you initially outlined. I'll definitely be needing help with the Metal part since I don't own any Apple Silicon, but I guess we'll cross that bridge when we get there.

**Reply by T3sT3ro** [reply]
What about adding a shadow effect to the background of split? For example inset shadow can be used to add a futuristic glow with blur radius, but it could also be used without blur to add the inner outline without affecting split colors at all. Not however that I have no idea how it works in GTK, I base it only on my web dev experience with HTML and CSS.

**Comment by mitchellh** [comment]
Another question: does this only apply while splits are active or does this apply to windows in general?

I think it makes sense to only apply with splits but just want to check what the behavior is of other programs.

**Reply by siaal** [reply]
In Kitty at least, the borders are only created across the edge against other splits, and there is no outside window border regardless of whether splits are active or not.
![image](https://github.com/user-attachments/assets/df13de34-697b-4e3b-b940-8d1030fde37e)
![image](https://github.com/user-attachments/assets/42fe7e9b-4033-4a94-88ac-738db3de1157)

**Reply by yetyetanotherusername** [reply]
Concur, that's also how I remember it working in tmux.

**Reply by T3sT3ro** [reply]
Personally I don't like the "split only" way of coloring, because it gets very confusing in a 2-pane split view. If the split has 1px only, then you can't sensibly tell at a glance which pane has focus when there is no focus around the whole pane next to the borders. In tmux they color half a border in that mode, and it is really hard to see depending on the color theme and remember which side of the split corresponds to which active pane.

<img width="1158" height="713" alt="image" src="https://github.com/user-attachments/assets/6e952863-4a59-4357-a7f6-8df6bd777b0e" />

**Reply by 8bittts** [reply]
i literally shared this just a moemnt ago :https://github.com/ghostty-org/ghostty/discussions/5806#discussioncomment-16073388

**Reply by 8bittts** [reply]
i think it's actually pretty helpful, but, not as useful if you just have two i guess.

**Reply by jcollie** [reply]
> Personally I don't like the "split only" way of coloring, because it gets very confusing in a vertical 2-pane split view. If the split has 1px only, then you can't sensibly tell at a glance which pane has focus when there is no focus around the whole pane next to the borders. In tmux they color half a border in that mode, and it is really hard to see depending on the color theme and remember which side of the split corresponds to which active pane.

This is a very old discussion, since this began we have added several options for making the splits that are not focused more obvious:

[unfocused-split-opacity](https://ghostty.org/docs/config/reference#unfocused-split-opacity)
[unfocused-split-fill](https://ghostty.org/docs/config/reference#unfocused-split-fill)
[split-divider-color](https://ghostty.org/docs/config/reference#split-divider-color)

You can also apply a custom shader, which has access to the focused state of the terminal:

[custom-shader](https://ghostty.org/docs/config/reference#split-divider-color)

If you're on Linux, you can use custom CSS to do a _lot_ (but see the discussion above):

[gtk-custom-css](https://ghostty.org/docs/config/reference#gtk-custom-css)

**Reply by yetyetanotherusername** [reply]
> i literally shared this just a moemnt ago :[#5806 (comment)](https://github.com/ghostty-org/ghostty/discussions/5806#discussioncomment-16073388)

Not what I originally had in mind, but I like it.

**Reply by yetyetanotherusername** [reply]
@jcollie How would one go about finding how to reference the active split in css, analogous to how 

```css
tabbar tabbox tab:selected
```
references the active tab symbol in the tab bar?

**Reply by jcollie** [reply]
Use the GTK inspector to dig around and use the appropriate [CSS selectors](https://docs.gtk.org/gtk4/css-overview.html#gtk-css-selectors) to find the `:active` terminal

**Reply by yetyetanotherusername** [reply]
Thanks for your guidance, unfortunately I'm getting stuck at the following:

I tried

```css
paned overlay:active
```
which visually lets me outline the border correctly, but that only seems to trigger while i hold mouse button down over a split, it does nothing depending on whether a split is selected or not. Also tried `:focused`.

**Reply by yetyetanotherusername** [reply]
Trial and error to the rescue. Posting this here in case it is helpful for anyone. This draws a white border around the active split. Many thanks to jcollie for the guidance.

```css
paned box
{
    border-style: solid;
    border-color: #1c1c1f;
    border-width: 1px;
}

paned box:focus-within
{
    border-style: solid;
    border-color: #ffffff;
    border-width: 1px;
}
```

**Reply by yetyetanotherusername** [reply]
Spoke a bit to soon, as it seems that after a period of inactivity, the `focus-visible` property is lost and only returns after typing/changing split/etc.

Apparently I'm not selecting the correct element with `paned box`, which is why `:focus` seems to fail.

`paned box:focus-within` works as expected, though.

**Comment by yetyetanotherusername** [comment]
I have found a [solution](https://github.com/ghostty-org/ghostty/discussions/2727#discussioncomment-16085499) that _I'm_ very happy with drawing a colored border around the active split. 

I'm a bit hesitant marking this as answered though since the solution is not _exactly_ what I originally proposed in this issue and it is gtk-only. I guess it makes sense keeping it open until someone implements a solution working on Metal as well?

**Comment by gkoreli** [comment]
I've been building a custom theme with a focus shader and hit the fundamental limitation: shaders render *inside* the terminal surface and can never touch the divider. The divider is the literal boundary between splits — it should be the primary focus signal.

**Proposal:**

```ini
focused-split-divider-color = #4cffb2
unfocused-split-divider-color = #191926
```

Backward compatible — if unset, falls back to `split-divider-color`.

To answer @mitchellh's [question from Jan 2025](https://github.com/ghostty-org/ghostty/discussions/2727#discussioncomment-11712490): this applies only when splits exist. Single-pane windows have no divider to color.

**Implementation is smaller than people think:**

**macOS (~30 lines of SwiftUI):**
- [`TerminalSplitTreeView.swift:72`](https://github.com/ghostty-org/ghostty/blob/main/macos/Sources/Features/Splits/TerminalSplitTreeView.swift#L72) — currently passes a single `ghostty.config.splitDividerColor` to all dividers
- [`SplitView.swift:13`](https://github.com/ghostty-org/ghostty/blob/main/macos/Sources/Features/Splits/SplitView.swift#L13) — stores `let dividerColor: Color`
- Fix: pass both colors + a binding to which child is focused. The split tree already tracks focus state.

**GTK (CSS class toggle):**
- [`surface.blp:206`](https://github.com/ghostty-org/ghostty/blob/main/src/apprt/gtk/ui/1.2/surface.blp#L206) already binds `template.focused`
- [`surface.zig:2708`](https://github.com/ghostty-org/ghostty/blob/main/src/apprt/gtk/class/surface.zig#L2708) already fires `updateFocus()` on focus change
- Fix: toggle a CSS class on the parent paned when a child gains focus. @yetyetanotherusername's [workaround](https://github.com/ghostty-org/ghostty/discussions/2727#discussioncomment-16085499) (`paned box:focus-within`) proves the CSS approach works — it just needs to be wired up natively.

**Config definition:**
- [`Config.zig:1086`](https://github.com/ghostty-org/ghostty/blob/main/src/config/Config.zig#L1086) — add two new options right after `split-divider-color`, same `?Color` type.

**This intentionally covers color only** — thickness, animation, and effects are the domain of [custom shaders](https://ghostty.org/docs/config/reference#custom-shader), which already have full GPU access inside the terminal surface. The divider answers "which split am I in?" — the shader answers "how fancy do you want it to look?"

**Prior art:**
- kitty: [`active_border_color`](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.active_border_color) / [`inactive_border_color`](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.inactive_border_color)
- tmux: [`pane-active-border-style`](https://man.openbsd.org/OpenBSD-current/man1/tmux.1#pane-active-border-style) / [`pane-border-style`](https://man.openbsd.org/OpenBSD-current/man1/tmux.1#pane-border-style)

**Why it matters:**
- Zero performance cost (color swap on focus change, no compositing)
- Cross-platform with minimal code
- Complements shaders and `unfocused-split-opacity` rather than replacing them
- Unblocks theme authors who want focus identity without dimming

**Comment by LeszekBlazewski** [comment]
This would be a lovely feature!
